File: frontend/src/optical/kinds/waveplate/physics.ts
/**
 * Waveplate PhysicsOp — birefringent retarder.
 *
 * Op name: `jones_waveplate`
 * Kind:    `waveplate`
 *
 * Jones matrix in beam-local s/p basis with fast axis at angle θ from +s:
 *
 *   J(δ, θ) = R(-θ) · diag(1, e^(iδ)) · R(θ)
 *
 * Where:
 *   - δ = retardance in radians (HWP: π, QWP: π/2)
 *   - θ = fast-axis angle from +s (degrees in params)
 *   - R(α) = [[cos α, sin α], [-sin α, cos α]]  (this is our `rotateJones`'s
 *            internal matrix — see rotateJones doc)
 *
 * Implementation step-by-step:
 *   1. Express Jones in fast/slow basis: rotateJones(jones, θ)
 *   2. Apply phase delay to slow axis: E_sl *= e^(iδ)
 *   3. Express back in s/p basis: rotateJones(result, -θ)
 *
 * Power conservation: unitary transform, so |jones_out|² = |jones_in|².
 * powerMw passes through unchanged.
 *
 * Required params:
 *   - retardanceDeg (default 90 = QWP). HWP: 180.
 *   - fastAxisDegBeamLocal (default 0)
 *
 * Optional: lengthMm + refractiveIndex for q-parameter slab propagation.
 */

import {
  type BeamRay,
  vec3Add,
  vec3Distance,
  vec3Scale,
} from "../../beam-ray";
import { type Complex, cAdd, cDiv, cMul } from "../../fiber/gaussian";
import { rotateJones } from "../../jones";
import {
  type PhysicsOp,
  type PhysicsOpContext,
  registerKind,
} from "../../registry";

function applyAbcdToQ(A: number, B: number, C: number, D: number, q: Complex): Complex {
  const Aq: Complex = { re: A * q.re, im: A * q.im };
  const Bc: Complex = { re: B, im: 0 };
  const Cq: Complex = { re: C * q.re, im: C * q.im };
  const Dc: Complex = { re: D, im: 0 };
  return cDiv(cAdd(Aq, Bc), cAdd(Cq, Dc));
}

/** Apply waveplate Jones matrix to (E_s, E_p) with retardance δ and
 *  fast axis at θ from +s. */
export function applyWaveplate(
  jones: [Complex, Complex],
  retardanceRad: number,
  fastAxisRad: number,
): [Complex, Complex] {
  // 1. Rotate s/p → fast/slow basis
  const [Ef, Esl] = rotateJones(jones, fastAxisRad);

  // 2. Apply phase e^(iδ) to slow axis
  const phaseSlow: Complex = {
    re: Math.cos(retardanceRad),
    im: Math.sin(retardanceRad),
  };
  const EslShifted = cMul(phaseSlow, Esl);

  // 3. Rotate fast/slow → s/p basis
  return rotateJones([Ef, EslShifted], -fastAxisRad);
}

export const jonesWaveplateOp: PhysicsOp = (
  rayIn: BeamRay,
  ctx: PhysicsOpContext,
): BeamRay[] => {
  const retDeg = (ctx.params.retardanceDeg as number | undefined) ?? 90;
  const retardanceRad = (retDeg * Math.PI) / 180;
  const fastDeg = (
    ctx.params.fastAxisDegBeamLocal
    ?? ctx.params.fastAxisDegBodyLocal
    ?? 0
  ) as number;
  const fastAxisRad = (fastDeg * Math.PI) / 180;

  const newJones = applyWaveplate(rayIn.jones, retardanceRad, fastAxisRad);

  // Slab q propagation (q' = q + B with B = L/n).
  const L = (ctx.params.lengthMm ?? ctx.params.thicknessMm) as number | undefined;
  const n = (ctx.params.refractiveIndex as number | undefined) ?? 1.0;
  const B = (typeof L === "number" && L > 0)
    ? L / n
    : vec3Distance(ctx.faceIn.positionMmBodyLocal, ctx.faceOut.positionMmBodyLocal);

  const qxOut = applyAbcdToQ(1, B, 0, 1, rayIn.qx);
  const qyOut = applyAbcdToQ(1, B, 0, 1, rayIn.qy);

  // Chief ray exits at face_out position. Direction unchanged.
  const thickness = vec3Distance(
    ctx.faceIn.positionMmBodyLocal,
    ctx.faceOut.positionMmBodyLocal,
  );
  const newOrigin = vec3Add(rayIn.origin, vec3Scale(rayIn.direction, thickness));

  return [{
    ...rayIn,
    origin: newOrigin,
    jones: newJones,
    qx: qxOut,
    qy: qyOut,
    pathLengthMm: rayIn.pathLengthMm + thickness,
    // power conserved (unitary)
  }];
};

registerKind("waveplate", {
  ops: {
    jones_waveplate: jonesWaveplateOp,
  },
  needsAperture: true,
});

File: backend/app/optical/kinds/waveplate/physics.py
"""Waveplate PhysicsOp (Python mirror of frontend/src/optical/kinds/waveplate/physics.ts).

Op name: jones_waveplate
Kind: waveplate
Params:
  - retardanceDeg (default 90 = QWP). HWP: 180.
  - fastAxisDegBeamLocal (default 0)
  - lengthMm, refractiveIndex (for slab q propagation)

Jones matrix: J(δ, θ) = R(-θ) · diag(1, e^(iδ)) · R(θ)
"""

from __future__ import annotations

import math

from app.optical.beam_ray import BeamRay, vec3_distance
from app.optical.jones import rotate_jones
from app.optical.registry import (
    KindEntry,
    PhysicsOpContext,
    register_kind,
)


def _apply_abcd_to_q(A: float, B: float, C: float, D: float, q: complex) -> complex:
    return (A * q + B) / (C * q + D)


def apply_waveplate(
    jones: tuple[complex, complex],
    retardance_rad: float,
    fast_axis_rad: float,
) -> tuple[complex, complex]:
    """J(δ, θ) = R(-θ) · diag(1, e^iδ) · R(θ) applied to (E_s, E_p)."""
    # 1. Rotate s/p → fast/slow
    e_f, e_sl = rotate_jones(jones, fast_axis_rad)
    # 2. Phase delay on slow axis
    phase_slow = complex(math.cos(retardance_rad), math.sin(retardance_rad))
    e_sl_shifted = phase_slow * e_sl
    # 3. Rotate fast/slow → s/p
    return rotate_jones((e_f, e_sl_shifted), -fast_axis_rad)


def jones_waveplate_op(ray_in: BeamRay, ctx: PhysicsOpContext) -> list[BeamRay]:
    ret_deg = float(ctx.params.get("retardanceDeg", 90.0))
    retardance_rad = math.radians(ret_deg)
    fast_deg = float(
        ctx.params.get("fastAxisDegBeamLocal")
        or ctx.params.get("fastAxisDegBodyLocal")
        or 0.0
    )
    fast_axis_rad = math.radians(fast_deg)

    new_jones = apply_waveplate(ray_in.jones, retardance_rad, fast_axis_rad)

    L = ctx.params.get("lengthMm", ctx.params.get("thicknessMm"))
    n = float(ctx.params.get("refractiveIndex", 1.0))
    if isinstance(L, (int, float)) and L > 0:
        B = float(L) / n
    else:
        B = vec3_distance(
            ctx.face_in.position_mm_body_local,
            ctx.face_out.position_mm_body_local,
        )

    qx_out = _apply_abcd_to_q(1, B, 0, 1, ray_in.qx)
    qy_out = _apply_abcd_to_q(1, B, 0, 1, ray_in.qy)

    thickness = vec3_distance(
        ctx.face_in.position_mm_body_local,
        ctx.face_out.position_mm_body_local,
    )
    new_origin = ray_in.origin + ray_in.direction * thickness

    return [ray_in.replaced(
        origin=new_origin,
        jones=new_jones,
        qx=qx_out,
        qy=qy_out,
        path_length_mm=ray_in.path_length_mm + thickness,
    )]


register_kind("waveplate", KindEntry(
    ops={"jones_waveplate": jones_waveplate_op},
    needs_aperture=True,
))
